### ISSUE TITLE
Cannot assign from [MaybeNull] to [AllowNull]

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#38926]
**Version Used**: 89b4f60648fadeebdd53088fe8391bf9197bba7b

**Steps to Reproduce**:

[Sharplab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWAChGABAJgEYWAGC59gOgAiASwgBzXAHtYwgMZR+AYUko4AQRqEAnlGFQA3IxYBmXqwqKAPABUAfBQDejCi4oBtALIQtwOADkiQiQKNWJJAHcA4gBdCmsKAA9grUMmBldeU2YAFgoPAAoASkdnDNcEigBeChTSlwBfRka0oA==)

```csharp
#nullable enable
#pragma warning disable 649

using System;
using System.Diagnostics.CodeAnalysis;

public class C<T> {
    [MaybeNull, AllowNull] T x, y;

    public void M() {
        x = y; // This should not report a warning
    }
}
```

**Expected Behavior**:

No warnings.

**Actual Behavior**:

1. CS8717: A member returning a [MaybeNull] value introduces a null value when 'T' is a non-nullable reference type (this issue)
1. Two warnings CS8618 (separate issue #37511)

**Comment by gafter** [comment]
This would be addressed by https://github.com/dotnet/roslyn/issues/38638
